src/slurm_plugin/clustermgtd.py

```diff
@@ -904,6 +1030,111 @@ def _terminate_orphaned_instances(self, cluster_instances, ips_used_by_slurm):
+
+    @staticmethod
+    def _get_partitions_to_disable(nodename_to_slurm_nodes_map, bootstrap_failure_partitions, partitions):
+        """Get partitions not running jobs to disable."""
+        # Get partitions have jobs running
+        partitions_have_active_jobs = ClusterManager._filter_partitions_with_active_jobs(
+            nodename_to_slurm_nodes_map, partitions
+        )
+        # Only disable bootstrap failure partitions that not having active jobs running
+        partitions_to_disable = set(bootstrap_failure_partitions) - partitions_have_active_jobs
+        bootstrap_failure_partitions_have_jobs = set(bootstrap_failure_partitions) - partitions_to_disable
+        if bootstrap_failure_partitions_have_jobs:
+            log.info(
+                f"Bootstrap failure partitions {bootstrap_failure_partitions_have_jobs} currently have jobs running, "
+                "not disabling them."
+            )
+        return partitions_to_disable
+
+    def _is_protected_mode_enabled(self):
+        """When protected_failure_count is set to -1, disable protected mode."""
+        if self._config.protected_failure_count == -1:
+            return False
+        return True
+
+    def _handle_protected_mode_process(self, slurm_nodes, partitions, bootstrap_failure_nodes, healthy_nodes):
+        """Handle the process of entering protected mode."""
+        # Handle successfully launched nodes
+        nodename_to_slurm_nodes_map = {node.name: node for node in slurm_nodes}
+        self._handle_successfully_launched_nodes(healthy_nodes)
+
+        # Handle bootstrap failures nodes
+        if bootstrap_failure_nodes:
+            log.warning("Found the following bootstrap failure nodes: %s", print_with_count(bootstrap_failure_nodes))
+            self._handle_bootstrap_failure_nodes(bootstrap_failure_nodes)
+
+        # Enter protected mode
+        # We will put a partition into inactive state only if the partition satisfies the following:
+        # Partition is not INACTIVE
+        # Partition bootstrap failure count above threshold
+        # Partition does not have job running
+        logging.info(f"Partitions_protected_failure_count_map is {self._partitions_protected_failure_count_map}")
+        partitions_above_threshold = [
+            part
+            for part, failures in self._partitions_protected_failure_count_map.items()
+            if part not in self._inactive_partitions and failures >= self._config.protected_failure_count
+        ]
+        if partitions_above_threshold:
+            log.info(f"Bootstrap failure partitions: {partitions_above_threshold}")
+            partitions_to_disable = ClusterManager._get_partitions_to_disable(
+                nodename_to_slurm_nodes_map, partitions_above_threshold, partitions
+            )
+            self._enter_protected_mode(partitions_to_disable)
+
+    @staticmethod
+    def _filter_partitions_with_active_jobs(nodename_to_slurm_nodes_map, partitions):
```

Comment:
Dropping this comment here but it is more of a general consideration: the ClusterManger class is becoming too complex and implementing functionalities that could be encapsulated in separate classes. For example it could make sense to have a Partition class which includes as an attribute the list of nodes and that exposes a set of functionalities such has the has_jobs so that you could simply do partition.has_jobs(). The same applies to Node related operations where you could actually have 2 concrete classes for dynamic and static nodes which will allow you to remove all those if/else statements in the code. 
Other than simplifying the code this would allow to reduce the number of times we have to transform, copy or pas s around all the various data structures the ClusterManager is currently using and that have a linear dependency on the number of nodes. Ideally we could identify the following entities: Scheduler, Partition, Node (StaticNode, DynamicNode), EC2Instance that have the following dependencies:
* Scheduler->List[Partition], Scheduler->List[Nodes]
* Partition->List[Nodes]
* Node->EC2Instance

Now I can't asking to refactor the code as part of this PR to achieve all of this but if we could think of some reasonable enhancement to start extracting some of the functionalities outside of the ClusterManager I think it is the right move.

Reply:
Will address refactoring in future PR, resolving for now